File: public/templates/dnf.java
import java.util.Arrays;

public class dnf {
    //sort the array of 0 ,1 and 2 using dutch national flag algo

    static void sort(int a[], int arr_size)
    {
        int lo = 0;
        int hi = arr_size - 1;
        int mid = 0, temp = 0;

        while (mid <= hi) {


        switch (a[mid]) {
            // If the element is 0
            case 0 -> {
                temp = a[lo];
                a[lo] = a[mid];
                a[mid] = temp;
                lo++;
                mid++;
                break;
            }
            case 1 -> mid++;
            case 2 -> {
                temp = a[mid];
                a[mid] = a[hi];
                a[hi] = temp;
                hi--;
                break;
            }
        }
        }
    }
    public static void main(String[] args){


        int [] arr = new int[]{0,1,1,0,2,1,1,1,2,2,2,0,2,1};
        int n = arr.length;

        sort(arr, n);
        System.out.println(Arrays.toString(arr));

    }
}

File: public/templates/dnf.py
# invarient 0...lo-1 all zeros, lo...mid-1 all 1 and mid to hi all 2
# O(n)
def dnf(A):
    hi = len(A)-1
    lo = 0
    mid = 0
    while(mid<hi):
        if A[mid]==0:
            while lo<len(A)-1 and A[lo]==0: lo +=1 
            A[lo],A[mid] = A[mid],A[lo]  
        elif A[mid]==2:
            while hi>0 and A[hi]==2: hi -=1 
            A[hi],A[mid] = A[mid],A[hi]  
        else: mid +=1

def main():
    A = [1,0,1,0,2,1,0,2]
    dnf(A)
    print(A)